src/js/components/modals/ServiceFormModal.js

```diff
@@ -0,0 +1,244 @@
+import mixin from 'reactjs-mixin';
+import {Modal} from 'reactjs-components';
+import React from 'react';
+import {StoreMixin} from 'mesosphere-shared-reactjs';
+import Ace from 'react-ace';
+
+import 'brace/mode/json';
+import 'brace/theme/monokai';
+import 'brace/ext/language_tools';
+
+import AdvancedConfig from '../AdvancedConfig';
+import MarathonStore from '../../stores/MarathonStore';
+import Service from '../../structs/Service';
+import ServiceUtil from '../../utils/ServiceUtil';
+import ServiceSchema from '../../constants/ServiceSchema';
+import ToggleButton from '../ToggleButton';
+
+const METHODS_TO_BIND = [
+  'getAdvancedSubmit',
+  'handleChange',
+  'handleCancel',
+  'handleClearError',
+  'handleJSONChange',
+  'handleJSONToggle',
+  'handleSubmit',
+  'onMarathonStoreServiceCreateError',
+  'onMarathonStoreServiceCreateSuccess'
+];
+
+class ServiceFormModal extends mixin(StoreMixin) {
+  constructor() {
+    super(...arguments);
+
+    let model =
+      ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.state = {
+      service: ServiceUtil.createServiceFromFormModel(model),
+      model: model,
+      json: false,
+      app: JSON.stringify({id:'', cmd:''}, null, 2),
+      errorMessage: null
+    };
+
+    this.store_listeners = [
+      {
+        name: 'marathon',
+        events: ['serviceCreateError', 'serviceCreateSuccess']
+      }
+    ];
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    super.componentWillReceiveProps(...arguments);
+    let {props} = this;
+    if (!props.open && nextProps.open) {
+      this.resetState();
+    }
+  }
+
+  resetState() {
+    let model = ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.setState({
+      service: ServiceUtil.createServiceFromFormModel(model),
+      model: model,
+      app: '',
+      errorMessage: null
+    });
+  }
+
+  handleChange({isValidated, model}) {
+    if (isValidated) {
+      this.setState({
+        service: ServiceUtil.createServiceFromFormModel(model),
+        model: model,
+        errorMessage: null
+      });
+    }
+  }
+
+  handleClearError() {
+    this.setState({
+      errorMessage: null
+    });
+  }
+
+  handleJSONChange(app) {
+    let {service} = this.state;
+    try {
+      service = new Service(JSON.parse(app));
+    } catch (e) {
+
+    }
+    this.setState({app, service})
+  }
+
+  handleJSONToggle() {
+    if (!this.state.json) {
+      let model = this.triggerAdvancedSubmit().model;
+      let service = ServiceUtil.createServiceFromFormModel(model)
+      this.setState({
+        app: JSON.stringify(ServiceUtil
+          .getAppDefinitionFromService(service), null, 2),
+        model,
+        service
+      });
+    }
+    this.setState({json: !this.state.json});
```

Comment:
if `!this.state.json` is true, then we'd be setting state twice in a row. i think react might be smart enough to do it in one go, but could we just create an object called `nextState` and then have one `setState` at the end?